query.go

```diff
@@ -177,62 +177,112 @@ func argsFromTokens(tokens []token) ([]string, error) {
-				write = true
-
-				if len(f.Fields) > 0 {
-					return false, fmt.Errorf("field %s set to true in sparse fieldset map has children fields, needs submap for children fields", f.Decl.Name)
-				}
+		if len(f.Fields) == 0 {
+			return false, fmt.Errorf("field %s set to a submap of fields in sparse fieldset map has no children fields, needs to be set to true or false", f.Decl.Name)
+		}
+	default:
+		// Include case when fields equals nil
+		write = false
+	}
+
+	if f.Keep {
+		write = true
+	}
+
+	if !write {
+		return false, nil
+	}
+
+	f.Decl.tokenize(w)
+	for _, directive := range f.Directives {
+		io.WriteString(w, " ") //nolint:errcheck
+		directive.tokenize(w)
+	}
+
+	if len(f.Fields) > 0 {
+		io.WriteString(w, " {\n") //nolint:errcheck
+
+		for _, field := range f.Fields {
+			var written bool
+			var err error
+
+			switch ts := fields.(type) {
+			case Fields:
+				written, err = field.tokenizeWithFields(w, ts[field.Decl.Name])
+			default:
+				written, err = field.tokenizeWithFields(w, nil)
 			}
-		case Fields:
-			if len(f.Fields) == 0 {
-				return false, fmt.Errorf("field %s set to a submap of fields in sparse fieldset map has no children fields, needs to be set to true or false", f.Decl.Name)
+
+			if err != nil {
+				return false, err
 			}
 
-			write = true
-			fields = ts
-		default:
-			write = false
+			if written {
+				io.WriteString(w, "\n") //nolint:errcheck
+			}
 		}
+		io.WriteString(w, "}") //nolint:errcheck
 	}
 
-	if write {
-		f.Decl.tokenize(w)
-		for _, directive := range f.Directives {
-			io.WriteString(w, " ") //nolint:errcheck
-			directive.tokenize(w)
-		}
+	return write, nil
+}
 
-		if len(f.Fields) > 0 {
-			io.WriteString(w, " {\n") //nolint:errcheck
-			for _, field := range f.Fields {
-				written, err := field.tokenize(w, fields)
-				if err != nil {
-					return false, err
-				}
+// tokenize recurses through a field to write all of the information contained
+// within the root field as well as all of it's children field to any type that
+// implements the io.Writer interface.
+//
+// Returns a bool denoting whether or not the field was written and an error.
+func (f *field) tokenize(w io.Writer, fields Fields) (bool, error) { //nolint:gocyclo
+	f.Decl.tokenize(w)
+	for _, directive := range f.Directives {
+		io.WriteString(w, " ") //nolint:errcheck
+		directive.tokenize(w)
+	}
 
-				if written {
-					io.WriteString(w, "\n") //nolint:errcheck
-				}
+	var written bool
+	var err error
+
+	if len(f.Fields) > 0 {
+		io.WriteString(w, " {\n") //nolint:errcheck
+
+		for _, field := range f.Fields {
+			if fields == nil {
+				written, err = field.tokenize(w, nil)
+			} else {
+				written, err = field.tokenizeWithFields(w, fields)
+			}
+
+			if err != nil {
+				return false, err
+			}
+
+			if written {
+				io.WriteString(w, "\n") //nolint:errcheck
 			}
-			io.WriteString(w, "}") //nolint:errcheck
 		}
+		io.WriteString(w, "}") //nolint:errcheck
 	}
 
-	return write, nil
+	return true, nil
```

Comment:
should this be written?

Reply:
`written` is used to know the outcome of children. For the current field we already know it was written since its name and directives (lines 255 and 258) were added to the buffer.